### PR TITLE
Add optional column_order in JSON reader

### LINKED ISSUE #16799: [BUG] `cudf::io::read_json` does not verify output column structures with the input schema
Look at this example:
```
  std::string const json_string = R"({"data": [1, 2, 3], "data": [4, 5, 6]})";

  std::map<std::string, cudf::io::schema_element> dtype_schema;
  dtype_schema["data"] = cudf::io::schema_element{cudf::data_type{cudf::type_id::STRUCT}, {}};

  auto& child_schema = dtype_schema["data"].child_types;
  child_schema["a"]  = cudf::io::schema_element{cudf::data_type{cudf::type_id::INT64}, {}};
  child_schema["b"]  = cudf::io::schema_element{cudf::data_type{cudf::type_id::INT64}, {}};

  cudf::io::json_reader_options json_lines_options =
    cudf::io::json_reader_options::builder(
      cudf::io::source_info{json_string.c_str(), json_string.size()})
      .dtypes(dtype_schema)
      .lines(true);

  auto const table = cudf::io::read_json(json_lines_options);
  cudf::test::print(table.tbl->get_column(0).view());
```
Output:
```
cudf::list_view<int64_t>:
Length : 1
Offsets : 0, 3
   1, 2, 3, 4, 5, 6
```

Notice that the input schema in the example above is `"data": STRUCT<"a": INT64, "b": INT64>`. On the other hand, the `data` field in the input JSON is an array instead. We can observe that the `data` column after parsing (which is a lists column) is just gathered for generating the output. The desired output here should be all nulls instead, and the correct behavior here should be to make sure the output columns have the right types/structures that we are looking for as specified in the input schema.

### LINKED ISSUE #17090: [FEA] The output columns of `read_json` need to follow depth-first-search order as in the input schema
Currently, `read_json` takes in the read schema as a map. As such, its column order cannot be preserved when the output table is generated. The callers need to keep track of the column order by themselves and rearrange the columns based on the output column names.

We should support preserving the column order (by depth-first-search order) so the callers can reduce their overhead of doing so. In order to do so, the input schema needs to be specified using a `std::vector` of nested columns instead of `std::map`.

### LINKED ISSUE #17091: [FEA] `read_json` should output all-nulls columns for the schema columns that do not exist in the input
In order to fulfill some applications such as Spark, `read_json` needs to output all-nulls columns for the columns in the input schema that do not exist in the input data. This applies to all schema columns at any nested level.

If there is conflict of interest between applications, we can implement this as a reader option.

### LINKED ISSUE #17240: [BUG] JSON parser still returns outdated schema structure for strings column
Previously, strings columns have two children: `offsets` and `chars`. As such, JSON parsers return this schema:
```
      // For string columns return ["offsets", "char"] schema
      if (target_type.id() == type_id::STRING) {
        return {std::move(col), std::vector<column_name_info>{{"offsets"}, {"chars"}}};
      }
```

However, recent changes in cudf have removed the `chars` child column, leaving only the `offsets` child remaining. We should update JSON parser to return the proper schema.

### PR DESCRIPTION
**Pull request by karthikeyann**
## Description
This PR adds optional column order to enforce column order in the output. This feature is required by spark from_json. 

Optional `column_order` is added to `schema_element`, and it is validated during reader_option creation. The column order can be specified at root level and for any struct in any level.
• For root level, the dtypes should be schema_element with type STRUCT. (schema_element is added to variant dtypes)
• For nested level, column_order can be specified for any STRUCT type. (could be a map of schema_element , or schema_element)
If the column order is not specified, the order of columns is same as the order of columns that appear in json file.

Closes #17240 (metadata updated) 
Closes #17091 (will return all nulls column if not present in input json)
Closes #17090 (fixed with new schema_element as dtype)
Closes #16799 (output columns are created from column_order if present)

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cudf/io/json.hpp [hunk 1]
```diff
@@ -53,6 +53,69 @@ struct schema_element {
...
+  std::optional<std::vector<std::string>> column_order;
+
+  /**
+   * @brief Default constructor
+   *
+   * @param type The type that this column should be converted to
+   * @param child_types Allows specifying this column's child columns target type
+   * @param column_order Allows specifying the order of the columns
+   */
+  schema_element(data_type type,
```
[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/include/cudf/io/json.hpp:70]: I tried this. It still had ambiguous overload that I reported. 
```
In file included from /home/coder/cudf/cpp/tests/io/json/json_test.cpp:29:
/home/coder/cudf/cpp/include/cudf/io/json.hpp:665:32: note: candidate: 'cudf::io::json_reader_options_builder& cudf::io::json_reader_options_builder::dtypes(std::map<std::__cxx11::basic_string<char>, cudf::data_type>)'
  665 |   json_reader_options_builder& dtypes(std::map<std::string, data_type> types)
      |                                ^~~~~~
/home/coder/cudf/cpp/include/cudf/io/json.hpp:677:32: note: candidate: 'cudf::io::json_reader_options_builder& cudf::io::json_reader_options_builder::dtypes(std::map<std::__cxx11::basic_string<char>, cudf::io::schema_element>)'
  677 |   json_reader_options_builder& dtypes(std::map<std::string, schema_element> types)
```

File: cpp/include/cudf/io/json.hpp [hunk 2]
```diff
@@ -53,6 +53,69 @@ struct schema_element {
...
+  {
+  }
+
+  /**
+   * @brief Constructor to create a schema_element from a data_type
+   *
+   * @param type The type that this column should be converted to
+   */
+  schema_element(data_type type) : schema_element{type, {}, std::nullopt} {}
+  schema_element() = default;
```
[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/include/cudf/io/json.hpp:85]: In places where std::map<string, schema_element> dmap;
`dmap["a"] = {data_type{STRING}}` is causing error because `dmap["a"] ` calls default constructor.
I removed all the constructors and kept the interface as simple struct.

File: cpp/include/cudf/io/json.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (lamarrr) [reply cpp/include/cudf/io/json.hpp:70]: did you remove the other constructors?

[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/include/cudf/io/json.hpp]: yes. I removed them.

**Comment by karthikeyann** [comment]
@ttnghia This PR is ready for testing. This will enforce the column order and also insert empty all-null columns if not present in the json data.

File: cpp/include/cudf/io/json.hpp [hunk 3]
```diff
@@ -53,6 +53,16 @@ struct schema_element {
...
+
+  /** @brief Allows specifying the order of the columns
+   */
+  std::optional<std::vector<std::string>> column_order;
+
+  /** @brief Returns the size of the child_types map
+   *
+   * @return size_t The size of the child_types map
+   */
+  size_t size() const { return child_types.size(); }
```
Comment (ttnghia) [inline cpp/include/cudf/io/json.hpp:65]: `size` is ambiguous here, since it means number of children. Should be `num_children()` instead.

File: cpp/include/cudf/io/json.hpp [hunk 4]
```diff
@@ -92,7 +102,8 @@ class json_reader_options {
   // Data types of the column; empty to infer dtypes
   std::variant<std::vector<data_type>,
                std::map<std::string, data_type>,
-               std::map<std::string, schema_element>>
+               std::map<std::string, schema_element>,
+               schema_element>
```
Comment (ttnghia) [inline cpp/include/cudf/io/json.hpp:106]: This is repeated in many places. To reduce maintenance effort and mistake, let's define them as a type alias:
```
using dtype_variant = std::variant<...>;
dtype_variant _dtypes;
```

File: cpp/src/io/json/host_tree_algorithms.cu [hunk 5]
```diff
@@ -596,6 +597,15 @@ std::pair<cudf::detail::host_vector<bool>, hashmap_of_device_columns> build_tree
...
                        mark_is_pruned(first_child_id, user_dtypes.at(name));
                    }
+                 },
+                 [&root_list_col_id, &adj, &mark_is_pruned, &column_names](
+                   schema_element const& user_dtypes) -> void {
+                   for (size_t i = 0; i < adj[root_list_col_id].size(); i++) {
+                     auto const first_child_id = adj[root_list_col_id][i];
+                     auto name                 = column_names[first_child_id];
+                     if (user_dtypes.child_types.count(name))
+                       mark_is_pruned(first_child_id, user_dtypes.child_types.at(name));
```
Comment (ttnghia) [inline cpp/src/io/json/host_tree_algorithms.cu:607]:
```suggestion
                     if (user_dtypes.child_types.count(name)) {
                       mark_is_pruned(first_child_id, user_dtypes.child_types.at(name));
                     }
```

File: cpp/src/io/json/json_column.cu [hunk 6]
```diff
@@ -410,12 +410,31 @@ std::pair<std::unique_ptr<column>, std::vector<column_name_info>> device_json_co
...
-        column_names.emplace_back(col->first);
-        auto& child_col           = col->second;
+      auto const& col_order = schema.has_value() and schema.value().column_order.has_value() and
+                                  not schema.value().column_order->empty()
+                                ? schema.value().column_order.value()
+                                : json_col.column_order;
+      for (auto const& col_name : col_order) {
         auto child_schema_element = get_child_schema(col_name);
+        auto const& found_col     = json_col.child_columns.find(col_name);
+        if (prune_columns and found_col == std::end(json_col.child_columns)) {
```
Comment (ttnghia) [inline cpp/src/io/json/json_column.cu:425]: `find` returns an rvalue iterator thus we should not bind it to a reference.
```suggestion
        auto const found_it     = json_col.child_columns.find(col_name);
        if (prune_columns and found_it == std::end(json_col.child_columns)) {
```

File: cpp/src/io/json/json_column.cu [hunk 7]
```diff
@@ -410,12 +410,31 @@ std::pair<std::unique_ptr<column>, std::vector<column_name_info>> device_json_co
...
+          column_names.emplace_back(make_column_name_info(
+            child_schema_element.value_or(schema_element{data_type{type_id::EMPTY}}), col_name));
+          auto all_null_column = make_all_nulls_column(
+            child_schema_element.value_or(schema_element{data_type{type_id::EMPTY}}),
```
Comment (ttnghia) [inline cpp/src/io/json/json_column.cu:429]: Here we should have the schema available. If not, there should be something wrong and we should throw instead of using `EMPTY` type.

File: cpp/src/io/json/json_column.cu [hunk 8]
```diff
@@ -410,12 +410,31 @@ std::pair<std::unique_ptr<column>, std::vector<column_name_info>> device_json_co
...
+            num_rows,
+            stream,
+            mr);
+          child_columns.emplace_back(std::move(all_null_column));
+          continue;
+        }
+        column_names.emplace_back(found_col->first);
```
Comment (ttnghia) [inline cpp/src/io/json/json_column.cu:436]: Above we have `if (prune_columns and found_col == std::end` thus here if `!prune_columns` then we still have `found_col == std::end`.

File: cpp/src/io/json/json_column.cu [hunk 9]
```diff
@@ -601,6 +635,11 @@ table_with_metadata device_parse_nested_json(device_span<SymbolT const> d_input,
           return (user_dtypes.find(col_name) != std::end(user_dtypes))
                    ? user_dtypes.find(col_name)->second
                    : std::optional<schema_element>{};
+        },
+        [col_name](schema_element const& user_dtypes) -> std::optional<schema_element> {
+          return (user_dtypes.child_types.find(col_name) != std::end(user_dtypes.child_types))
+                   ? user_dtypes.child_types.find(col_name)->second
```
Comment (ttnghia) [inline cpp/src/io/json/json_column.cu:641]: `find(col_name)` should not be called twice.

File: cpp/src/io/json/json_column.cu [hunk 10]
```diff
@@ -601,6 +635,11 @@ table_with_metadata device_parse_nested_json(device_span<SymbolT const> d_input,
...
+                   : std::optional<schema_element>{};
```
Comment (ttnghia) [inline cpp/src/io/json/json_column.cu:642]: `std::nullopt`?

File: cpp/src/io/json/json_column.cu [hunk 11]
```diff
@@ -624,6 +663,23 @@ table_with_metadata device_parse_nested_json(device_span<SymbolT const> d_input,
     debug_schema_print(child_schema_element);
 #endif
 
+    auto found_col = root_struct_col.child_columns.find(col_name);
```
Comment (ttnghia) [inline cpp/src/io/json/json_column.cu:666]:
```suggestion
    auto const found_it = root_struct_col.child_columns.find(col_name);
```

File: cpp/include/cudf/io/json.hpp (hunk 3 above)
Comment (karthikeyann) [reply cpp/include/cudf/io/json.hpp:65]: This is added to allow std::visit function work. It's not used anywhere else.

File: cpp/include/cudf/io/json.hpp (hunk 4 above)
Comment (karthikeyann) [reply cpp/include/cudf/io/json.hpp:106]: only at 2 places. Replaced it.

File: cpp/src/io/json/json_column.cu (hunk 7 above)
Comment (karthikeyann) [reply cpp/src/io/json/json_column.cu:429]: I added CUDF_EXPECTS.

File: cpp/src/io/json/json_column.cu (hunk 8 above)
Comment (karthikeyann) [reply cpp/src/io/json/json_column.cu:436]: I added `prune_columns` condition to `col_order` decision. That case won't happen.

File: cpp/src/io/json/json_column.cu (hunk 9 above)
Comment (karthikeyann) [reply cpp/src/io/json/json_column.cu:639]: Great suggestion!

**Comment by ttnghia** [comment]
Please add error check for the case that the input schema is specified by `std::map` but it also has the column order. In such cases, we need to throw.

**Comment by ttnghia** [comment]
Tested with spark-rapids (requires https://github.com/rapidsai/cudf/pull/17180) and all integration tests pass.

**Comment by karthikeyann** [comment]
Current validation of dtypes input schema.
-  If `column_order` and `child_types` size is not same, it throws cudf::logic_error
-  If names in `column_order` is not present in `child_types`, then it throws cudf::logic_error

@ttnghia could you clarify the requirement for dtypes as `schema_element` with column order?

- should column_order be present always?
- if column_order is not present, it follows the order of columns that's present in the file. is that desirable or should error be thrown?

File: cpp/include/cudf/io/json.hpp [hunk 12]
```diff
@@ -53,6 +54,10 @@ struct schema_element {
    * @brief Allows specifying this column's child columns target type
    */
   std::map<std::string, schema_element> child_types;
+
+  /** @brief Allows specifying the order of the columns
+   */
```
Comment (PointKernel) [inline cpp/include/cudf/io/json.hpp:59]:
```suggestion
  /** 
   * @brief Allows specifying the order of the columns
   */
```

File: cpp/src/io/json/parser_features.cpp [hunk 13]
```diff
@@ -16,14 +16,194 @@
...
+}  // namespace cudf::io
+
 namespace cudf::io::json::detail {
 
+/// Created an empty column of the specified schema
+struct empty_column_functor {
+  rmm::cuda_stream_view stream;
+  rmm::device_async_resource_ref mr;
+
+  template <typename T, std::enable_if_t<!cudf::is_nested<T>()>* = nullptr>
```
Comment (PointKernel) [inline cpp/src/io/json/parser_features.cpp:76]: could probably use `CUDF_ENABLE_IF`

File: cpp/include/cudf/io/json.hpp [hunk 14]
```diff
@@ -390,6 +394,16 @@ class json_reader_options {
    */
   void set_dtypes(std::map<std::string, schema_element> types) { _dtypes = std::move(types); }
 
+  /**
+   * @brief Set data types for a potentially nested column hierarchy.
+   *
+   * @throw cudf::logic_error if the `types` does not have column order specified.
```
Comment (ttnghia) [inline cpp/include/cudf/io/json.hpp:400]:
```suggestion
```

File: cpp/include/cudf/io/json.hpp [hunk 15]
```diff
@@ -390,6 +394,16 @@ class json_reader_options {
...
+   *
```
Comment (ttnghia) [inline cpp/include/cudf/io/json.hpp:401]:
```suggestion
```

File: cpp/src/io/json/parser_features.cpp [hunk 16]
```diff
@@ -16,14 +16,194 @@
...
+      schema.child_types.at(child_name).type, *this, schema.child_types.at(child_name));
+    auto offsets = make_empty_column(data_type(type_to_id<size_type>()));
+    return make_lists_column(0, std::move(offsets), std::move(child), 0, {}, stream, mr);
+  }
+
+  template <typename T, std::enable_if_t<std::is_same_v<T, cudf::struct_view>>* = nullptr>
+  std::unique_ptr<column> operator()(schema_element const& schema) const
+  {
+    std::vector<std::unique_ptr<column>> child_columns;
+    for (auto child_name : schema.column_order.value_or(std::vector<std::string>{})) {
```
Comment (PointKernel) [inline cpp/src/io/json/parser_features.cpp:97]:
```suggestion
    for (auto const& child_name : schema.column_order.value_or(std::vector<std::string>{})) {
```

File: cpp/src/io/json/json_column.cu [hunk 17]
```diff
@@ -410,12 +410,38 @@ std::pair<std::unique_ptr<column>, std::vector<column_name_info>> device_json_co
       std::vector<std::unique_ptr<column>> child_columns;
       std::vector<column_name_info> column_names{};
       size_type num_rows{json_col.num_rows};
+      if (schema.has_value() and schema.value().column_order.has_value()) {
+        CUDF_EXPECTS(schema.value().child_types.size() == schema.value().column_order->size(),
+                     "Input schema column order size mismatch with input schema child types");
+      }
```
Comment (ttnghia) [inline cpp/src/io/json/json_column.cu:416]: Use `validate_column_order`?

Comment (ttnghia) [reply cpp/src/io/json/json_column.cu:416]: Or probably we can remove this completely, since the schema has already been validated when it was set.

File: cpp/src/io/json/parser_features.cpp [hunk 18]
```diff
@@ -16,14 +16,194 @@
...
+    return make_structs_column(0, std::move(child_columns), 0, {}, stream, mr);
+  }
+};
+
+/// Created all null column of the specified schema
+struct allnull_column_functor {
+  rmm::cuda_stream_view stream;
+  rmm::device_async_resource_ref mr;
+
+  auto make_zeroed_offsets(size_type size) const
```
Comment (PointKernel) [inline cpp/src/io/json/parser_features.cpp:110]: question: should we make this private?

File: cpp/src/io/json/nested_json.hpp [hunk 19]
```diff
@@ -429,6 +429,26 @@ table_with_metadata device_parse_nested_json(device_span<SymbolT const> input,
                                              rmm::cuda_stream_view stream,
                                              rmm::device_async_resource_ref mr);
 
+/** @brief Create all null column of a given nested schema
+ *
```
Comment (PointKernel) [inline cpp/src/io/json/nested_json.hpp:433]:
```suggestion
/**
 * @brief Create all null column of a given nested schema
```

File: cpp/src/io/json/nested_json.hpp [hunk 20]
```diff
@@ -429,6 +429,26 @@ table_with_metadata device_parse_nested_json(device_span<SymbolT const> input,
...
+ * @param mr resource with which to allocate
+ * @return The all null column
+ */
+std::unique_ptr<column> make_all_nulls_column(schema_element const& schema,
+                                              size_type num_rows,
+                                              rmm::cuda_stream_view stream,
+                                              rmm::device_async_resource_ref mr);
+
+/** @brief Create metadata for a column of a given schema
+ *
```
Comment (PointKernel) [inline cpp/src/io/json/nested_json.hpp:446]:
```suggestion
/**
 * @brief Create metadata for a column of a given schema
```

File: cpp/src/io/json/json_column.cu [hunk 21]
```diff
@@ -576,11 +602,24 @@ table_with_metadata device_parse_nested_json(device_span<SymbolT const> d_input,
...
 
-  // Iterate over the struct's child columns and convert to cudf column
-  size_type column_index = 0;
-  for (auto const& col_name : root_struct_col.column_order) {
-    auto& json_col = root_struct_col.child_columns.find(col_name)->second;
+  bool const has_column_order =
+    options.is_enabled_prune_columns() and
+    std::holds_alternative<schema_element>(options.get_dtypes()) and
+    std::get<schema_element>(options.get_dtypes()).column_order.has_value() and
+    not std::get<schema_element>(options.get_dtypes()).column_order->empty();
```
Comment (ttnghia) [inline cpp/src/io/json/json_column.cu:609]: Can we throw if `options.is_enabled_prune_columns()` is `false` but column order is given? The users may want to specify column order but they may forget to enable prune column.

File: cpp/src/io/json/parser_features.cpp [hunk 22]
```diff
@@ -16,14 +16,194 @@
...
+                                              rmm::device_async_resource_ref mr)
+{
+  return cudf::type_dispatcher(schema.type, allnull_column_functor{stream, mr}, schema, num_rows);
+}
+
+column_name_info make_column_name_info(schema_element const& schema, std::string const& col_name)
+{
+  column_name_info info;
+  info.name = col_name;
+  if (schema.type.id() == type_id::STRUCT) {
```
Comment (PointKernel) [inline cpp/src/io/json/parser_features.cpp:186]: It's not a strict requirement, but I personally find that using a switch statement provides a cleaner solution in this situation compared to if-else statements.

File: cpp/src/io/json/parser_features.cpp [hunk 23]
```diff
@@ -16,14 +16,194 @@
...
+          if (!validate_column_order(it->second)) { return false; }
+        }
+      }
+    }
+  } else if (types.type.id() == type_id::LIST) {
+    if (types.child_types.size() != 1) { return false; }
+    auto it = types.child_types.begin();
+    if (it->second.type.id() == type_id::STRUCT or it->second.type.id() == type_id::LIST) {
+      if (!validate_column_order(it->second)) { return false; }
+    }
```
Comment (ttnghia) [inline cpp/src/io/json/parser_features.cpp:56]: Here we still didn't check for column name existence.

Comment (ttnghia) [reply cpp/src/io/json/parser_features.cpp:56]: Should be
```
  auto it = types.child_types.find(column_name);
  if (it == types.child_types.end()) { return false; }
```

File: cpp/src/io/json/parser_features.cpp [hunk 24]
```diff
@@ -16,14 +16,194 @@
...
+namespace cudf::io {
+namespace {
+bool validate_column_order(schema_element const& types)
+{
+  // For struct types, check if column_order size matches child_types size and all elements in
+  // column_order are in child_types, in child_types, call this function recursively.
+  // For list types, check if child_types size is 1 and call this function recursively.
+  if (types.type.id() == type_id::STRUCT) {
+    if (types.column_order.has_value()) {
+      if (types.column_order.value().size() != types.child_types.size()) { return false; }
```
Comment (ttnghia) [inline cpp/src/io/json/parser_features.cpp:42]: This line should be moved to the very beginning of the function.

File: cpp/src/io/json/parser_features.cpp [hunk 25]
```diff
@@ -16,14 +16,194 @@
...
+      info.children.push_back(make_column_name_info(child_schema, child_name));
+    }
+  } else if (schema.type.id() == type_id::DICTIONARY32) {
+    info.children.emplace_back("indices");
+    for (auto const& [child_name, child_schema] : schema.child_types) {
+      info.children.push_back(make_column_name_info(child_schema, child_name));
+    }
+  } else if (schema.type.id() == type_id::STRING) {
+    info.children.emplace_back("offsets");
+    info.children.emplace_back("chars");
```
Comment (ttnghia) [inline cpp/src/io/json/parser_features.cpp:202]: Is this still correct? Now strings columns no longer have `chars` child.

Comment (ttnghia) [reply cpp/src/io/json/parser_features.cpp:202]: Filed https://github.com/rapidsai/cudf/issues/17240.

File: cpp/src/io/json/parser_features.cpp (hunk 23 above)
Comment (karthikeyann) [reply cpp/src/io/json/parser_features.cpp:56]: This is not required.  For non-nested type, there is no column_order.
Nested types are taken care by STRUCT, and LIST if cases.

File: cpp/src/io/json/json_column.cu (hunk 21 above)
Comment (karthikeyann) [reply cpp/src/io/json/json_column.cu:609]: It is possible, the user wants to prune column, but maintain the order as in json file. that's why that throw was not added.

File: cpp/src/io/json/parser_features.cpp (hunk 18 above)
Comment (karthikeyann) [reply cpp/src/io/json/parser_features.cpp:111]: done.

File: cpp/src/io/json/parser_features.cpp (hunk 22 above)
Comment (karthikeyann) [reply cpp/src/io/json/parser_features.cpp]: changed to switch statement

**Review by PointKernel — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]

File: cpp/tests/io/json/json_test.cpp [hunk 26]
```diff
@@ -2827,7 +2828,7 @@ TEST_F(JsonReaderTest, JSONMixedTypeChildren)
     EXPECT_EQ(result.metadata.schema_info[0].name, "Root");
     ASSERT_EQ(result.metadata.schema_info[0].children.size(), 1);
     EXPECT_EQ(result.metadata.schema_info[0].children[0].name, "Key");
-    ASSERT_EQ(result.metadata.schema_info[0].children[0].children.size(), 2);
+    ASSERT_EQ(result.metadata.schema_info[0].children[0].children.size(), 1);
```
Comment (vuule) [inline cpp/tests/io/json/json_test.cpp:2831]: why this change?

**Review by vuule — COMMENTED** [review]
partial review, nothing blocking

File: cpp/tests/io/json/json_test.cpp (hunk 26 above)
Comment (karthikeyann) [reply cpp/tests/io/json/json_test.cpp:2831]: To address https://github.com/rapidsai/cudf/issues/17240 
Strings column will not have `chars` as child anymore.

File: cpp/src/io/json/nested_json.hpp [hunk 27]
```diff
@@ -429,6 +429,28 @@ table_with_metadata device_parse_nested_json(device_span<SymbolT const> input,
...
+std::unique_ptr<column> make_all_nulls_column(schema_element const& schema,
+                                              size_type num_rows,
+                                              rmm::cuda_stream_view stream,
+                                              rmm::device_async_resource_ref mr);
+
+/**
+ * @brief Create metadata for a column of a given schema
+ *
+ * @param schema The schema of the column
+ * @param col_name The name of the column
```
[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/src/io/json/nested_json.hpp:450]:
```suggestion
 * @param col_name The name of the column
 * @return column metadata for a given schema
```

File: cpp/src/io/json/json_column.cu [hunk 28]
```diff
@@ -411,11 +411,33 @@ std::pair<std::unique_ptr<column>, std::vector<column_name_info>> device_json_co
       std::vector<column_name_info> column_names{};
       size_type num_rows{json_col.num_rows};
       // Create children columns
-      for (auto const& col_name : json_col.column_order) {
-        auto const& col = json_col.child_columns.find(col_name);
-        column_names.emplace_back(col->first);
-        auto& child_col           = col->second;
+      auto const& col_order = prune_columns and schema.has_value() and
```
Comment (vuule) [inline cpp/src/io/json/json_column.cu:414]: can we apply the pattern from line 601 here?

File: cpp/src/io/json/json_column.cu [hunk 29]
```diff
@@ -411,11 +411,33 @@ std::pair<std::unique_ptr<column>, std::vector<column_name_info>> device_json_co
...
+                                  schema.value().column_order.has_value() and
+                                  not schema.value().column_order->empty()
```
Comment (vuule) [inline cpp/src/io/json/json_column.cu:416]: I don't know if this works and/or is any better; nested optional checking looks so clunky to me.
```suggestion
      auto const& col_order = prune_columns and not schema.value_or({}).column_order.value_or({}).empty()
```

File: cpp/tests/io/json/json_test.cpp [hunk 30]
```diff
@@ -2991,4 +2992,264 @@ TEST_F(JsonReaderTest, LastRecordInvalid)
...
+    {"a": 1, "b": {}}
+    {"a": 1,                              "c": null}
+    )";
+  std::string json_string  = R"([
+    {"a": 1, "b": {"0": "abc", "1": [-1.]}, "c": true},
+    {"a": 1, "b": {"0": "abc"          }, "c": false},
+    {"a": 1, "b": {}},
+    {"a": 1,                              "c": null}
+    ])";
+  for (auto& [json_string, lines] : {std::pair{json_stringl, true}, {json_string, false}}) {
```
Comment (vuule) [inline cpp/tests/io/json/json_test.cpp:3010]: neat!

File: cpp/tests/io/json/json_test.cpp [hunk 31]
```diff
@@ -2991,4 +2992,264 @@ TEST_F(JsonReaderTest, LastRecordInvalid)
...
+      // Input schema column order size mismatch with input schema child types
+    }
+    //// different column name in order, Error
+    {
+      cudf::io::schema_element dtype_schema{data_type{cudf::type_id::STRUCT},
+                                            {
+                                              {"a", {dtype<int32_t>()}},
+                                            },
+                                            {{"b"}}};
+      EXPECT_THROW(in_options.set_dtypes(dtype_schema), cudf::logic_error);
```
Comment (vuule) [inline cpp/tests/io/json/json_test.cpp:3133]: We probably want to throw `std::invalid_argument` when we check the dtypes in the setter.

**Review by vuule — COMMENTED** [review]
did not expected such a large and robust feature, great stuff!
Got a several comments, mostly nits

File: cpp/src/io/json/host_tree_algorithms.cu [hunk 32]
```diff
@@ -593,6 +594,15 @@ std::pair<cudf::detail::host_vector<bool>, hashmap_of_device_columns> build_tree
...
```
[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/src/io/json/host_tree_algorithms.cu:604]: `std::map::operator[]` will [insert default value](https://en.cppreference.com/w/cpp/container/map/operator_at) if not present.

[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/json/host_tree_algorithms.cu:604]: Isn't is always present, since we check `user_dtypes.child_types.count(name)` ?

[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/src/io/json/host_tree_algorithms.cu:604]: During initial plan, I removed default constructor of `schema_element`. So, `std::map::operator[]` failed to compile.
Now, it's back. We could use either `.at` or `[]`.

File: cpp/tests/io/json/json_test.cpp (hunk 31 above)
Comment (karthikeyann) [reply cpp/tests/io/json/json_test.cpp:3133]: Replaced with `std::invalid_argument`

File: cpp/src/io/json/json_column.cu (hunk 29 above)
Comment (karthikeyann) [reply cpp/src/io/json/json_column.cu:416]: `value_or({})` does not work. so, 
The code will be like
`not schema.value_or(schema_element{}).column_order.value_or(std::vector<std::string>{}).empty();`

**Review by ttnghia — APPROVED** [review]

**Review by vuule — APPROVED** [review]
thank you for addressing all suggestions!
